src/Cdn77/Sniffs/NamingConventions/ValidVariableNameSniff.php

```diff
@@ -0,0 +1,171 @@
+
+                    if (! $this->matchesRegex($objVarName, $this->memberPattern)) {
+                        $error = sprintf('Member variable "%%s" does not match pattern "%s"', $this->memberPattern);
+                        $data = [$objVarName];
+                        $phpcsFile->addError($error, $var, self::CODE_MEMBER_DOES_NOT_MATCH_PATTERN, $data);
+                    }
+                }
+            }
+        }
+
+        $objOperator = $phpcsFile->findPrevious(T_WHITESPACE, $stackPtr - 1, null, true);
+        if ($tokens[$objOperator]['code'] === T_DOUBLE_COLON) {
+            if (! $this->matchesRegex($varName, $this->memberPattern)) {
+                $error = sprintf('Member variable "%%s" does not match pattern "%s"', $this->memberPattern);
+                $data = [$tokens[$stackPtr]['content']];
+                $phpcsFile->addError($error, $stackPtr, self::CODE_MEMBER_DOES_NOT_MATCH_PATTERN, $data);
+            }
+
+            return;
+        }
+
+        if ($this->matchesRegex($varName, $this->pattern)) {
+            return;
+        }
+
+        $error = sprintf('Variable "%%s" does not match pattern "%s"', $this->pattern);
+        $data = [$varName];
+        $phpcsFile->addError($error, $stackPtr, self::CODE_DOES_NOT_MATCH_PATTERN, $data);
+    }
+
+    /**
+     * Processes class member variables.
+     *
+     * @param File $phpcsFile The file being scanned.
+     * @param int  $stackPtr  The position of the current token in the stack passed in $tokens.
+     *
+     * @phpcsSuppress SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingNativeTypeHint
+     */
+    protected function processMemberVar(File $phpcsFile, $stackPtr): void
+    {
+        $tokens = $phpcsFile->getTokens();
+
+        $varName = ltrim($tokens[$stackPtr]['content'], '$');
+        $memberProps = $phpcsFile->getMemberProperties($stackPtr);
+        if ($memberProps === []) {
+            // Couldn't get any info about this variable, which
+            // generally means it is invalid or possibly has a parse
+            // error. Any errors will be reported by the core, so
+            // we can ignore it.
+            return;
+        }
+
+        $errorData = [$varName];
+
+        if ($this->matchesRegex($varName, $this->memberPattern)) {
+            return;
+        }
+
+        $error = sprintf('Member variable "%%s" does not match pattern "%s"', $this->memberPattern);
+        $phpcsFile->addError($error, $stackPtr, self::CODE_MEMBER_DOES_NOT_MATCH_PATTERN, $errorData);
+    }
+
+    /**
+     * Processes the variable found within a double quoted string.
+     *
+     * @param File $phpcsFile The file being scanned.
+     * @param int  $stackPtr  The position of the double quoted string.
+     *
+     * @phpcsSuppress SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingNativeTypeHint
+     */
+    protected function processVariableInString(File $phpcsFile, $stackPtr): void
```

Comment:
will it work for herodoc?

Reply:
Have you checked tests?

Reply:
yop. But there no test for invalid var name in herodoc.

it works for:
```
$foo = <<<BAR
{$var_name}
BAR;
```
but not for: 
```
$foo = <<<BAR
$var_name
BAR;
```

Reply:
Got it now. 

I think it's fine as the sniff reports invalid var declarations successfully. No need to report usages. IMO. WDYT?

Reply:
I guess its ok.